feat(links): Add support for various links available with a snap

soumyaDghosh opened the pull request.

soumyaDghosh commented:
@robert-ancell @sergio-costas

sergio-costas commented:
@soumyaDghosh Could you add a description indicating in a more specific way what this patch does, please? Which links are added, for example.

soumyaDghosh commented:
Yes, sorry I didn't add it in the first place.

soumyaDghosh commented:
@sergio-costas Will it be possible for you to join the Snapcraft clinic of tomorrow if you want to discuss on something related to this PR.

soumyaDghosh commented:
@sergio-costas Wait a min, I just fixed the formatting. Haven't fixed the tests yet.

sergio-costas commented:
For the formatting, please, run the `format-source.sh` script. It will fix the format. Also, I strongly recommend you to add it to `.git/hooks/pre-commit` as a check, just by setting it to:

```
#!/bin/sh

./format-source.sh check
```

That will ensure that the format is right before uploading it.

sergio-costas commented:
Oh, ok. Sorry.

sergio-costas commented:
BTW, have you signed the Canonical contributor agreement? https://ubuntu.com/legal/contributors

soumyaDghosh commented:
> BTW, have you signed the Canonical contributor agreement? https://ubuntu.com/legal/contributors

Yes almost an year ago😁

soumyaDghosh commented:
@sergio-costas I have updated the tests. But there is a segfault until I print the values [here](https://github.com/canonical/snapd-glib/pull/207/files#diff-77b120fa6978a73634e4a48e1ec4d24d9dc30f260c900e7d38ab0311f08798eaR81-R83). The examples work perfectly. I ran gdb and there was some place with null value in the string or something probably, somewhere in the json builder or MockLink in the test, I am not handling memory properly.

soumyaDghosh commented:
@sergio-costas please rerun the CI, I have fixed the tests.

soumyaDghosh commented:
@sergio-costas do we need to keep supporting 20.04? GStrvBuilder is the correct way to create GStrv now, it was introduced in glib 2.68

https://docs.gtk.org/glib/struct.StrvBuilder.html

soumyaDghosh commented:
Okay, @sergio-costas fixed for 20.04 too. But I am not sure if creating GStrv like this is the best way.

sergio-costas commented:
> @sergio-costas do we need to keep supporting 20.04? 

Yes, I'm afraid of. But what you can do is to add some `#if GLIB_CHECK_VERSION`, and a comment to specify for which Ubuntu version is that required. That would allow to remove old code when those versions become unsupported.

sergio-costas commented:
Also, ran the tests in your code with `valgrind` and there are two invalid accesses, which can be dangerous. They don't happen in `master`. Can you check it?

```
ok 122 /disconnect-interface/invalid
# End of disconnect-interface tests
# Start of find tests
==1726487== Conditional jump or move depends on uninitialised value(s)
==1726487==    at 0x49B259C: g_strdupv (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.8400.1)
==1726487==    by 0x48B290C: snapd_link_set_property (snapd-link.c:83)
==1726487==    by 0x4CCF962: ??? (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.8400.1)
==1726487==    by 0x4CD23DA: g_object_new_valist (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.8400.1)
==1726487==    by 0x4CD27CE: g_object_new (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.8400.1)
==1726487==    by 0x487FD5C: _snapd_json_parse_snap (snapd-json.c:1121)
==1726487==    by 0x48875B2: parse_get_find_response (snapd-get-find.c:156)
==1726487==    by 0x48A1692: parse_response (snapd-client.c:477)
==1726487==    by 0x48A23BA: read_cb (snapd-client.c:701)
==1726487==    by 0x4B6A090: ??? (in /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0.8400.1)
==1726487==    by 0x4982DE1: ??? (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.8400.1)
==1726487==    by 0x49F41F7: ??? (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.8400.1)
==1726487== 
==1726487== Conditional jump or move depends on uninitialised value(s)
==1726487==    at 0x49B25D5: g_strdupv (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.8400.1)
==1726487==    by 0x48B290C: snapd_link_set_property (snapd-link.c:83)
==1726487==    by 0x4CCF962: ??? (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.8400.1)
==1726487==    by 0x4CD23DA: g_object_new_valist (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.8400.1)
==1726487==    by 0x4CD27CE: g_object_new (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.8400.1)
==1726487==    by 0x487FD5C: _snapd_json_parse_snap (snapd-json.c:1121)
==1726487==    by 0x48875B2: parse_get_find_response (snapd-get-find.c:156)
==1726487==    by 0x48A1692: parse_response (snapd-client.c:477)
==1726487==    by 0x48A23BA: read_cb (snapd-client.c:701)
==1726487==    by 0x4B6A090: ??? (in /usr/lib/x86_64-linux-gnu/libgio-2.0.so.0.8400.1)
==1726487==    by 0x4982DE1: ??? (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.8400.1)
==1726487==    by 0x49F41F7: ??? (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.8400.1)
==1726487== 
ok 123 /find/query
```

soumyaDghosh commented:
> Also, ran the tests in your code with `valgrind` and there are two invalid accesses, which can be dangerous. They don't happen in `master`. Can you check it?

This is happening because I changed the way GStrvs for urls are created, but forgot to make them NULL terminated. Fixed it.

sergio-costas commented on tests/mock-snapd.c:
```diff
@@ -2468,6 +2489,24 @@ static JsonNode *make_snap_node(MockSnap *snap) {
+      MockLink *urls = link->data;
+      json_builder_set_member_name(builder, urls->type);
+      json_builder_begin_array(builder);
+      for (gint i = 0; urls->urls[i] != NULL; i++) {
```
Maybe this could be replaced by

```
for (gchar **url = urls->urls; *url != NULL; url++) {
    json_builder_add_string_value(builder, *url);
```

sergio-costas commented on tests/test-qt.cpp:
```diff
@@ -3641,6 +3646,17 @@ static void test_find_query() {
+  g_autoptr(GStrvBuilder) contact_url_builder = g_strv_builder_new();
+  g_strv_builder_add(contact_url_builder, "contact");
+  g_auto(GStrv) contact_urls = g_strv_builder_end(contact_url_builder);
+  mock_snap_add_link(s, "contact", contact_urls);
```
Same here

soumyaDghosh commented on tests/test-glib.c:
```diff
@@ -4221,6 +4226,17 @@ static void test_find_query(void) {
+  g_autoptr(GStrvBuilder) contact_url_builder = g_strv_builder_new();
+  g_strv_builder_add(contact_url_builder, "contact");
+  g_auto(GStrv) contact_urls = g_strv_builder_end(contact_url_builder);
+  mock_snap_add_link(s, "contact", contact_urls);
```
All Done!

sergio-costas reviewed (changes requested):
Sorry for the delay. Just some minor details.

sergio-costas commented on examples/snap-glib.c:
```diff
@@ -139,6 +138,21 @@ static int info (int argc, char **argv)
+        GPtrArray *links = snapd_snap_get_links(store_snap);
+        if (links != NULL && links->len > 0) {
+            g_printerr("links: |\n");
+            for     (guint i = 0; i < links->len; i++) {
```
I don't know why, but I have a sixth sense for extra spaces. In my previous job I drove my colleagues crazy every time they prepared a powerpoint for a client, because I spotted every extra blank space 😅

sergio-costas replied:
Anyway, clang-format didn't detect it, don't know why 🤔

soumyaDghosh commented on snapd-glib/requests/snapd-json.c:
```diff
@@ -1070,6 +1088,45 @@ SnapdSnap *_snapd_json_parse_snap(JsonNode *node, GError **error) {
+  JsonObject *links_object = _snapd_json_get_object(object, "links");
+  g_autoptr(GPtrArray) links_array =
+      g_ptr_array_new_with_free_func(g_object_unref);
+  GList *link_types = _snapd_json_get_keys(links_object);
```
Do you mean `g_autolist(GList)`? Because `g_autolist(gchar)` didn't work

soumyaDghosh replied:
Changing this to g_autolist was causing some memory leaks, so left it unchanged

soumyaDghosh replied:
Yes I did try `gchar *`. Do you mean passing `gchar *` from `_snapd_json_get_keys` itself?

sergio-costas replied:
Ok, it seems that it was already commented... https://gitlab.gnome.org/GNOME/glib/-/issues/1595 and the conclusion was to use a GPtrArray, which does support g_autolist.

sergio-costas replied:
I think it would be cleaner, yes. We should try to use automatic memory management as much as possible.

soumyaDghosh replied:
Now I remember why I made it a `GList`. `json_object_get_members` itself returns a GList, so changing it to `GPtrArray` will be a lot of extra changes.

sergio-costas replied:
But in `_snapd_json_get_keys()`  you still iterate over it to remove any NULL value, and put them into a new GList... You only have to replace there the `keys` GList with a `GPtrArray` and replace the `g_list_append()` call at line 80... Or am I missing something? (I need another coffee).

soumyaDghosh replied:
I am sending the patch tomorrow. Neon has broke `gobject-introspection` with update to `setuptools`, or may be it's broken in Noble too?

soumyaDghosh replied:
Yeah this issue has been patched in gobject-introspection 1.81.2

https://gitlab.gnome.org/GNOME/gobject-introspection/-/issues/?sort=created_date&state=closed&search=distutils&first_page_size=20&show=eyJpaWQiOiI1MTUiLCJmdWxsX3BhdGgiOiJHTk9NRS9nb2JqZWN0LWludHJvc3BlY3Rpb24iLCJpZCI6MjA1ODY5fQ%3D%3D

soumyaDghosh replied:
Done

sergio-costas commented on snapd-glib/requests/snapd-json.c:
```diff
@@ -67,6 +68,23 @@ const gchar *_snapd_json_get_string(JsonObject *object, const gchar *name,
+GList *_snapd_json_get_keys(JsonObject *object) {
+  GList *keys = NULL;
+  GList *members = NULL;
+  if (object != NULL) {
```
Mmm... I don't get it... if `object` is NULL, members will be NULL because it's never assigned; then, the `for` loop won't execute, and `keys` will be NULL too, so `g_steal_pointer()` will return NULL... which is what `g_return_val_if_fail` does... It doesn't halt the test, it's not a `g_assert_non_null`.

Or am I missing something...?

soumyaDghosh replied:
`g_return_val_if_fail(object != NULL, NULL);`

Atleast in this scenario, what it's doing is, if the object is `NULL`, it'll not go to the next step of loop, instead it'll return `NULL` from there, but it'll also a Critical Warning, this critical warning is breaking tests. Is this not how it should work?

sergio-costas replied:
Oh... you are right, didn't realize that those calls send a Critical Warning... it makes sense, then. Sorry.

What about just changing the `if` with the opposite?
```
    if (object == NULL) {
        return NULL;
    }
```

I think it makes the source a little bit cleaner. What do you think?

soumyaDghosh replied:
Done

sergio-costas commented on snapd-glib/requests/snapd-json.c:
```diff
@@ -1070,6 +1085,44 @@ SnapdSnap *_snapd_json_parse_snap(JsonNode *node, GError **error) {
+
+      g_autoptr(GPtrArray) urls_array = g_ptr_array_new();
+
+      JsonArray *values_array = _snapd_json_get_array(links_object, key);
```
`_snapd_json_get_array()` returns a new array, or increments the reference count of the array itself.
```suggestion
      g_autoptr(JsonArray) values_array = _snapd_json_get_array(links_object, key);
```

sergio-costas reviewed (changes requested):
A little memory leak. If I can merge it by myself, I'll do it.

sergio-costas reviewed (approved):
Everything looks great.

soumyaDghosh commented on snapd-glib/requests/snapd-json.c:
```diff
@@ -1070,6 +1085,44 @@ SnapdSnap *_snapd_json_parse_snap(JsonNode *node, GError **error) {
+
+      g_autoptr(GPtrArray) urls_array = g_ptr_array_new();
+
+      g_autoptr(JsonArray) values_array = _snapd_json_get_array(links_object, key);
```
clang-format didn't give me this formatting
